idx/memory/memory.go

```diff
@@ -1826,6 +1826,68 @@ ORGS:
 	return pruned, nil
 }
 
+// AddDef adds def to the index.
+func (m *UnpartitionedMemoryIdx) AddDefs(defs []schema.MetricDefinition) {
+	// TODO - refactor from AddOrUpdate
+	for _, def := range defs {
+		pre := time.Now()
+		m.RLockHigh()
+		existing, ok := m.defById[def.Id]
+		m.RUnlockHigh()
+		if ok {
+			updateExisting(existing, def.Partition, def.LastUpdate, pre)
+			continue
+		}
+		archive := createArchive(&def)
+		if m.writeQueue == nil {
+			// writeQueue not enabled, so acquire a wlock and immediately add to the index.
+			bc := m.Lock()
+			m.add(archive)
+			bc.Unlock("AddDefs", nil)
+			statAddDuration.Value(time.Since(pre))
+		} else {
+			// push the new archive into the writeQueue. If there is already an archive in the
+			// writeQueue with the same mkey, it will be replaced.
+			m.writeQueue.Queue(archive)
+		}
+	}
+}
+
+// DeleteDefs deletes the matching defs.
+func (m *UnpartitionedMemoryIdx) DeleteDefs(defs []schema.MetricDefinition, archive bool) {
+	// TODO - finer grained locking
+	bc := m.Lock()
+	defer bc.Unlock("DeleteDefs", nil)
+
+	taggedIds := make(map[uint32]IdSet)
+	for _, def := range defs {
+		if len(def.Tags) == 0 {
+			tree, ok := m.tree[def.OrgId]
+			if !ok {
+				continue
+			}
+
+			n, ok := tree.Items[def.Name]
+			if !ok || !n.Leaf() {
+				continue
+			}
+
+			// SEAN TODO - This seems like this could overdelete (if multiple defs for this path)
```

Comment:
you're referring to the findCache invalidation here right? interesting one. I suppose the only correct solution would be for the control server to include the original queries/path in the messages. and invalidate once per path. (similar to `UnpartitionedMemoryIdx.Delete()`)

the other thing we should do similar to that function is do the below instead of individual invalidations.
```
                if len(defs) > findCacheInvalidateQueueSize {
                        m.findCache.Purge(orgId)
                }
```

Reply:
It would be tough to add a top-level purge call as the OrgId is per def. As mentioned above, the control server doesn't currently issue deletes for untagged data, so this isn't an active issue. Should I add a TODO?

Reply:
sounds good, i'll leave this to you to handle (or simply add todo notes) however you like. that should cover this I think?